torch/testing/_internal/distributed/distributed_test.py

```diff
@@ -5778,13 +5778,29 @@ def test_broadcast_object_list(self):
                 else [None for _ in COLLECTIVES_OBJECT_TEST_LIST]
             )
 
-            # Single object test with device specified
-            single_obj_list = [objects[0]]
-            if self.rank != src_rank:
-                self.assertNotEqual(single_obj_list[0], COLLECTIVES_OBJECT_TEST_LIST[0])
-            device = torch.device('cpu') if backend == 'gloo' else torch.device(next_rank)
-            dist.broadcast_object_list(single_obj_list, src=0, group=None, device=device)
-            self.assertEqual(single_obj_list[0], COLLECTIVES_OBJECT_TEST_LIST[0])
+            # Single object test with device specified. Backend="gloo", device=cpu
+            if backend == "gloo":
+                single_obj_list = [objects[0]]
+                if self.rank != src_rank:
+                    self.assertNotEqual(single_obj_list[0], COLLECTIVES_OBJECT_TEST_LIST[0])
+                dist.broadcast_object_list(single_obj_list, src=0, group=None, device=torch.device('cpu'))
+                self.assertEqual(single_obj_list[0], COLLECTIVES_OBJECT_TEST_LIST[0])
+
+            # # Single object test with device specified. Backend="gloo", device=current_device+1
+            if backend == "gloo":
+                single_obj_list = [objects[0]]
+                if self.rank != src_rank:
+                    self.assertNotEqual(single_obj_list[0], COLLECTIVES_OBJECT_TEST_LIST[0])
+                dist.broadcast_object_list(single_obj_list, src=0, group=None, device=torch.device(next_rank))
```

Comment:
I mentioned previously:
> only run (2) if torch.cuda.device_count() == dist.get_world_size() to ensure that the line mentioned above will not error. 

I prefer to have some kind of check here to be defensive against varying test hardware configurations (e.g. if CI changes its setup, someone runs these tests locally with their own setup, etc.). This line with `device=torch.device(next_rank)` will error if there are no GPUs available. Currently, I think the testing ensures the invariant that if `backend="nccl"`, then there will be GPUs available, but I do not believe that that same guarantee is made for `backend="gloo"`. That is why I suggested to check that `torch.cuda.device_count() == dist.get_world_size()`. This guarantees that the scheme of setting the device to be `rank + 1` will not error. If you want to check the GPU availability in another way, feel free to do so.

Reply:
+1 to @andwgu's comment. Another option is extracting this test to its own unit test function, and add decorators like `@skip_if_lt_x()` accordingly.

https://github.com/pytorch/pytorch/blob/8a2c7d902fb159bb57a0cf46f3a5ba74095cc34a/torch/testing/_internal/distributed/distributed_test.py#L875

Reply:
@andwgu The test already has the following decorator - 
```
@require_n_gpus_for_nccl_backend(
            int(os.environ["WORLD_SIZE"]), os.environ["BACKEND"]
        )
```

shouldn't that be enough to ensure there are no errors here due to no GPUs available?

Reply:
Looks like this decorator only checks GPUs when the backend is NCCL, but not gloo?

https://github.com/pytorch/pytorch/blob/14f63763c16e77ac4793c938003e791dd71b217d/torch/testing/_internal/common_distributed.py#L79-L89

Reply:
Very great point! 
leveraged Andrew's suggestion here
if backend != "nccl" and torch.cuda.device_count() == int(self.world_size)